elide-core/src/main/java/com/yahoo/elide/core/EntityPermissions.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2015, Yahoo Inc.
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.core;
+
+import com.yahoo.elide.annotation.CreatePermission;
+import com.yahoo.elide.annotation.DeletePermission;
+import com.yahoo.elide.annotation.ReadPermission;
+import com.yahoo.elide.annotation.SharePermission;
+import com.yahoo.elide.annotation.UpdatePermission;
+import com.yahoo.elide.audit.InvalidSyntaxException;
+import com.yahoo.elide.generated.parsers.ExpressionLexer;
+import com.yahoo.elide.generated.parsers.ExpressionParser;
+import com.yahoo.elide.security.checks.Check;
+import lombok.extern.slf4j.Slf4j;
+import org.antlr.v4.runtime.ANTLRInputStream;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.misc.ParseCancellationException;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AccessibleObject;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Extract permissions related annotation data for a model.
+ */
+@Slf4j
+public class EntityPermissions {
+    private static final Class[] PERMISSION_ANNOTATIONS = new Class[]{
+            ReadPermission.class,
+            CreatePermission.class,
+            DeletePermission.class,
+            SharePermission.class,
+            UpdatePermission.class
+    };
+
+    public static final EntityPermissions EMPTY_PERMISSIONS = new EntityPermissions();
+
+    /**
+     * What type of permission.
+     */
+    public enum PermissionType {
+        EXPRESSION, ANY, ALL
+    }
+
+    /**
+     * Encapsulates the rules for a permission annotation.
+     */
+    public static class PermissionDefinition {
+        public final PermissionType permissionType;
+        public final Class<? extends Check>[] checks;
```

Comment:
I'm not sure I like this representation at all. We really should only have a single, definitive representation for checks. What this does is force a second code path below to construct an expression.

I propose we convert these checks to an expression and use the `ParseTree` as the definitive format since that is the long-term goal. I believe this should be straightforward:

```java
boolean anyChecks = ...; // Computed as usual
String conjunction = (anyChecks) ? " or " : " and ";
String expression = Arrays.asList(checks)
                               .stream()
                               .map(Class::getCanonicalName)
                               .reduce("", (left, right) -> (left.isEmpty()) ? right : left + conjunction + right);
```

You could then parse this string. 

This way we can guarantee everything is being constructed as intended for when we're ready to remove this functionality altogether.

I notice that this may seem like extra overhead, but since we're transitioning it will be logically more consistent and this should be short-lived pain. Since we plan on encouraging expression use anyway, we should not experience much pain here in the short-term either.